### PR TITLE
Allow DROP NOT NULL on column of compressed table

### LINKED ISSUE #5776: [Enhancement]: Remove columns not null on compressed tables
### What type of enhancement is this?

User experience

### What subsystems and features will be improved?

Compression

### What does the enhancement do?

It would be nice to be able to remove not null on a column in a compressed table without the need to decompress the whole table. For example I cant even do that, because I would run out of space.

### Implementation challenges

Unfortunatelly I dont have necessary knowledge to understand what would it take to implement and why compression prevents user to do that.

### PR DESCRIPTION
**Pull request by mkindahl**
Remove the restriction on `ALTER TABLE ... ALTER COLUMN SET/DROP NOT NULL` on compressed tables.

Dropping or setting the `NOT NULL` constraint on a column of a compressed table is safe since it does not require re-writing any data. Setting `NOT NULL` will require a full table scan to verify that there are no NULL in the table.

Fixes #5776

### CODE REVIEW

**Comment by mkindahl** [comment]
It is actually fine to use `SET NOT NULL` as well. It will generate a full table scan to investigate if there are any NULL present, but otherwise it should not require any dramatic changes.

File: tsl/test/expected/compression_ddl.out [hunk 1]
```diff
@@ -2536,3 +2536,47 @@ ALTER TABLE compression_drop DROP COLUMN v0;
...
+\set ON_ERROR_STOP 1
+ALTER TABLE test2 ALTER COLUMN i DROP NOT NULL;
+INSERT INTO test2(ts,b,t) VALUES ('2024-11-18 18:04:51',99,'magic');
+SELECT count(*) FROM test2 WHERE i IS NULL;
+ count 
+-------
+     1
+(1 row)
+
+SELECT count(compress_chunk(ch)) FROM show_chunks('test2') ch;
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply tsl/test/expected/compression_ddl.out:2570]: No, that is not tested. It tests that the inserted value (which ends up in the WS) can be found before and after compression, but it does not decompress the entire chunk.

**Review by erimatnor — APPROVED** [review]

**Review by antekresic — CHANGES_REQUESTED** [review]
Setting `NOT NULL` on a column which has compressed NULL values doesn't work properly. Maybe we should just limit this change to dropping `NOT NULL` since setting requires decompression which isn't done in the current change.

Also, lets add a test for this, trying to set `NOT NULL` on a column which has NULL values and the hypertable is full compressed (not partial).

**Comment by mkindahl** [comment]
> Setting `NOT NULL` on a column which has compressed NULL values doesn't work properly. Maybe we should just limit this change to dropping `NOT NULL` since setting requires decompression which isn't done in the current change.

Yeah, I discovered that it does not work for compressed tables using the `heap` access method, but this works fine for chunks with the `hypercore` access method (since it looks like a normal hypertable). Added checks so that we only allow this for compressed chunks that are using `hypercore` table access method.

> Also, lets add a test for this, trying to set `NOT NULL` on a column which has NULL values and the hypertable is full compressed (not partial).

There is an existing test in `compression_errors`, but that does not compress the table, so the question is if we should check the flag on the hypertable (and ignore the status of the chunks), or if we should allow this with the compression flag set and chunk fully decompressed.

**Comment by feikesteenbergen** [comment]
Could we get the simplest thing merged the `DROP NOT NULL`? the rest is also useful, but easier to workaround for users

**Comment by mkindahl** [comment]
> Could we get the simplest thing merged the `DROP NOT NULL`? the rest is also useful, but easier to workaround for users

Ok, split out the code to deal with `SET NOT NULL` into #7463 and left the other pull request here. Still awaiting reviews.

**Review by antekresic — APPROVED** [review]

**Comment by mkindahl** [comment]
> I guess my approval was for a different PR. I approve also this PR, but with some pause.
> 
> I am not 100% sure it is a good idea to allow dropping `NOT NULL` on compressed tables since it is not possible to add it back without decompressing. So, if people do it expecting that they can add it back, then they might be frustrated.

That is a good point, but it is hard to do something about it. We can print a warning when we remove it, but if people actually did not want to remove it, it is too late.
